### PR TITLE
test(unit): adds tests for Session Reconciler

### LINKED ISSUE #72: Unit test Session Reconciler
Look at the operator test framework

- [x] Separate object conversion logic out of session.go
- [ ] Separate between Object conversion/sync and server update

### PR DESCRIPTION
**Pull request by aslakknutsen**
Fixes #72

### CODE REVIEW

File: pkg/controller/session/session_controller_test.go [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+								{
+									Name:      "details",
+									Resources: []*v1alpha1.RefResource{{Kind: &kind, Name: &name, Action: &action}},
+								},
+							},
+						},
+					},
+				}
+			})
+			It("status updated", func() {
```
Comment (bartoszmajsak) [inline pkg/controller/session/session_controller_test.go]: How about making it a bit more elaborated - e.g. `It("should update status when .... ?"`

File: pkg/controller/session/session_controller_test.go [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+				locator.Action = foundTestLocator
+				mutator.Action = basicTestMutator(model.ResourceStatus{Name: "details2", Kind: "test", Action: model.ActionCreated})
+
+				res, err := controller.Reconcile(req)
+				Expect(err).ToNot(HaveOccurred())
+				Expect(res.Requeue).To(BeFalse())
+
+				Expect(locator.WasCalled).To(BeTrue())
+				Expect(mutator.WasCalled).To(BeTrue())
+				Expect(revertor.WasCalled).ToNot(BeTrue())
```
Comment (bartoszmajsak) [inline pkg/controller/session/session_controller_test.go]: Maybe these kind of verifications could actually be 3 separated tests - ensure we call locator, mutator but not revertor

File: pkg/controller/session/session_controller_test.go [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+					},
+				}
+			})
+			It("status updated", func() {
+				locator.Action = foundTestLocator
+				mutator.Action = basicTestMutator(model.ResourceStatus{Name: "details2", Kind: "test", Action: model.ActionCreated})
+
+				res, err := controller.Reconcile(req)
+				Expect(err).ToNot(HaveOccurred())
+				Expect(res.Requeue).To(BeFalse())
```
Comment (bartoszmajsak) [inline pkg/controller/session/session_controller_test.go]: Separated contract/test on its own?

File: pkg/controller/session/session_controller_test.go [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+
+				Expect(locator.WasCalled).To(BeTrue())
+				Expect(mutator.WasCalled).To(BeTrue())
+				Expect(revertor.WasCalled).ToNot(BeTrue())
+
+				modified := GetSession("test", "test-session")
+				Expect(modified.Status).ToNot(BeNil())
+				Expect(modified.Status.Refs).To(HaveLen(2))
+				Expect(modified.Status.Refs[0].Name).To(Equal("details"))
+				Expect(modified.Status.Refs[1].Name).To(Equal("details2"))
```
Comment (bartoszmajsak) [inline pkg/controller/session/session_controller_test.go]: Maybe we can transform refs extracting names and assert on slice instead? Not sure that improves readability much though... just a thought.

File: pkg/controller/session/session_controller_test.go [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+		controller                 reconcile.Reconciler
+		req                        reconcile.Request
+		schema                     *runtime.Scheme
+		c                          client.Client
+		locator, mutator, revertor = &trackedLocator{Action: notFoundTestLocator}, &trackedMutator{Action: emptyTestMutator}, &trackedRevertor{Action: emptyTestRevertor}
+	)
+	GetSession := func(c *client.Client) func(namespace, name string) v1alpha1.Session {
+		return func(namespace, name string) v1alpha1.Session {
+			s := v1alpha1.Session{}
+			err := (*c).Get(context.TODO(), types.NamespacedName{Namespace: namespace, Name: name}, &s)
```
Comment (bartoszmajsak) [inline pkg/controller/session/session_controller_test.go]: If we go with a generated client in #89 then we should redo this little thing

File: pkg/controller/session/model_convert_test.go [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+import (
+	"github.com/aslakknutsen/istio-workspace/pkg/apis/istio/v1alpha1"
+	"github.com/aslakknutsen/istio-workspace/pkg/controller/session"
+	"github.com/aslakknutsen/istio-workspace/pkg/model"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("Basic model convertion", func() {
```
Comment (bartoszmajsak) [inline pkg/controller/session/model_convert_test.go]: `var _ = Describe("Basic model conversion", func() {`

File: pkg/controller/session/model_convert_test.go [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+								Resources: []*v1alpha1.RefResource{
+									{
+										Kind: &kind, Name: &name, Action: &aFailed,
+									},
+								},
+							},
+						},
+					},
+				}
+			})
```
Comment (bartoszmajsak) [inline pkg/controller/session/model_convert_test.go]: image

Comment (bartoszmajsak) [reply pkg/controller/session/model_convert_test.go]: Makes me wonder if some builder wouldn't make it look a bit nicer overall :)

File: pkg/controller/session/model_convert.go [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+	refs := []*model.Ref{}
+	for _, statusRef := range session.Status.Refs {
+		r := &model.Ref{Name: statusRef.Name}
+		StatusToRef(session, r)
+		refs = append(refs, r)
+	}
+	return refs
+}
+
+// StatusToRef filles the ResourceStatus of a Ref based on the Session.Status.Refs with the same name
```
Comment (bartoszmajsak) [inline pkg/controller/session/model_convert.go]:
```suggestion
// StatusToRef fills the ResourceStatus of a Ref based on the Session.Status.Refs with the same name
```

File: pkg/controller/session/session_controller.go [hunk 9]
```diff
@@ -16,36 +16,40 @@ import (
...
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
-
 	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
-
 	"sigs.k8s.io/controller-runtime/pkg/source"
 )
 
 const (
-	finalizer = "finalizers.istio.workspace.session"
+	// Finalizer defins the Finalizer name owned by the Session reconciler
```
Comment (bartoszmajsak) [inline pkg/controller/session/session_controller.go]:
```suggestion
// Finalizer defines the Finalizer name owned by the Session reconciler
```

File: pkg/controller/session/session_controller_test.go (hunk 2 above)
Comment (aslakknutsen) [reply pkg/controller/session/session_controller_test.go]: I've split them up a little. Moved the WasCalled ones into it's own test. So there are two tests; Was what should be called called & Was the output the one expected.

https://github.com/aslakknutsen/istio-workspace/pull/90/commits/5c416866df4c6fc26d4058cba5ccf04ca2a64770

File: pkg/controller/session/session_controller_test.go (hunk 4 above)
Comment (aslakknutsen) [reply pkg/controller/session/session_controller_test.go]: Was wondering about it... but...

File: pkg/controller/session/session_controller_test.go (hunk 1 above)
Comment (aslakknutsen) [reply pkg/controller/session/session_controller_test.go]: https://github.com/aslakknutsen/istio-workspace/pull/90/commits/5c416866df4c6fc26d4058cba5ccf04ca2a64770

File: pkg/controller/session/model_convert_test.go (hunk 7 above)
Comment (aslakknutsen) [reply pkg/controller/session/model_convert_test.go]: No clue why GitHub is optimized for horizontal space. :man_shrugging: 

Screenshot from 2019-04-10 10-41-47

File: pkg/controller/session/session_controller_test.go (hunk 4 above)
Comment (bartoszmajsak) [reply pkg/controller/session/session_controller_test.go]: Yeah, let's leave that

**Pull request merged**